lib/resources/oidc.js

```diff
@@ -0,0 +1,292 @@
+
+      log('code_verifier:', code_verifier);
+
+      const code_challenge = generators.codeChallenge(code_verifier);
+
+      const authUrl = client.authorizationUrl({
+        scope: SCOPES.join(' '),
+        resource: `${envDomain}/v1`,
+        code_challenge,
+        code_challenge_method: CODE_CHALLENGE_METHOD,
+      });
+
+      res.cookie(CODE_VERIFIER_COOKIE, code_verifier, { ...callbackCookieProps, maxAge:ONE_HOUR });
+
+      const { next } = req.query;
+      if(next) res.cookie(NEXT_COOKIE, next, { ...callbackCookieProps, maxAge:ONE_HOUR });
+
+      res.redirect(authUrl); // looks like this would break some expectation of endpoint() managing headers
+    } catch(err) {
+      return serverError(res, err);
+    }
+  });
+
+  // FIXME we need to use endpoint() to give access to Audits,Sessions,Users, but there seems to be something
+  // it expects to be responsible for in the response lifecycle, causing:
+  //
+  //   Error: The resource returned no data. This is likely a developer problem.
+  //
+  // Still, things seem to work anyway.
+  service.get('/oidc/callback', endpoint(async (container, {}, req, res) => {
+    try {
+      const code_verifier = req.cookies[CODE_VERIFIER_COOKIE];
+      const next          = req.cookies[NEXT_COOKIE];
+      res.clearCookie(CODE_VERIFIER_COOKIE, callbackCookieProps);
+      res.clearCookie(NEXT_COOKIE,          callbackCookieProps);
+
+      log('code_verifier:', code_verifier);
+
+      const client = await getClient();
+
+      const params = client.callbackParams(req);
+      const tokenSet = await client.callback(getRedirectUri(), params, { code_verifier });
+      log('received and validated tokens:', tokenSet);
+      log('validated ID Token claims:', tokenSet.claims())
+
+      const { access_token } = tokenSet;
+
+      const userinfo = await client.userinfo(access_token);
+
+      const { email, email_verified } = userinfo;
+      if(!email) {
+        container.Sentry.captureException(new Error(`Required claim not provided in UserInfo Response: 'email'`));
+        return errorToFrontend(req, res, 'provider-misconfigured');
+      }
+      if(!email_verified) return errorToFrontend(req, res, 'email-not-verified');
+
+      log('userinfo:', userinfo);
+
+      const user = await getUserByEmail(container, email);
+      if(!user) return errorToFrontend(req, res, 'auth-ok-user-not-found');
+
+      const session = await initSession(container, req, res, user);
+
+      if(DEBUG_IDP_RESPONSE) {
+        return debugIdpResponse(res, session, userinfo, user);
+      }
+
+      // This redirect is neat, but breaks SameSite: Secure cookies.
+      //return redirect(302, '/'); // REVIEW: internally, the redirect() function throws... which is odd
+      // Instead, we need to render a page and then "browse" from that page to the normal frontend:
+      // TODO id=cl only set for playwright.. why can't it locate this anchor in any other way?
+
+      const nextPath = safeNextPathFrom(next);
```

Comment:
If we add a generic redirect route in Frontend, maybe we could remove `safeNextPathFrom()` and use a path like this:

```js
const nextPath = `/#/redirect?next=${next}`;
```

Though maybe `next` needs to be encoded for use as a query parameter.

Reply:
There's one complication here that I hadn't thought about, which is redirects to Enketo. If you navigate directly to an Enketo form that requires cookie auth, but you aren't logged into Central, then you're redirected to the Frontend login page, with `next` set to the Enketo form. Once you log into Central, Frontend sends you back to Enketo (in the `AccountLogin` component).

Right now, `safeNextPathFrom()` seems to always redirect the user to Frontend, which means that it doesn't work if `next` is to Enketo. If we replaced `safeNextPathFrom()` with a Frontend `/#/redirect` path, then we could have Frontend continue to handle all the logic around redirecting to Enketo. That feels like the easiest thing from a complexity standpoint: it's one less thing for Backend to do, and there'd be no code duplication between Frontend and Backend. Right now, I don't think Backend makes an assumptions about where Enketo is deployed. In contrast, Frontend assumes that Enketo is deployed at `/-`. It seems better for Backend not to be making a similar assumption and to continue to hand off to Frontend the responsibility of handling the `next` query parameter.

That said, I do think there's a performance cost to that choice. If `next` is to an Enketo form, then Backend will redirect to Frontend, which will redirect to Enketo. In other words, there's more redirecting happening than what's strictly necessary. All things considered, I lean toward having Frontend handle all `next` redirects, but I can see it both ways.

Reply:
I've added an e2e test for enketo redirects, and implemented.